File: Android/app/src/main/java/leapfrog_inc/icchi/Http/Requester/ItemRequester.java
package leapfrog_inc.icchi.Http.Requester;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

import leapfrog_inc.icchi.Function.Constants;
import leapfrog_inc.icchi.Http.HttpManager;

/**
 * Created by Leapfrog-Software on 2018/01/27.
 */

public class ItemRequester {

    public static class ItemData {

        public String itemId;
        public String name;
        public String kana;

        static public ItemData create(JSONObject json) {

            try {
                String itemId = json.getString("itemId");
                String name = json.getString("name");
                String kana = json.getString("kana");

                ItemData itemData = new ItemData();
                itemData.itemId = itemId;
                itemData.name = name;
                itemData.kana = kana;

                return itemData;

            } catch(Exception e) {}

            return null;
        }
    }

    private static ItemRequester requester = new ItemRequester();

    private ItemRequester(){}

    public static ItemRequester getInstance(){
        return requester;
    }

    private ArrayList<ItemData> mDataList = new ArrayList<ItemData>();

    public void fetch(final ItemRequesterCallback callback){

        HttpManager httpManager = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {
                if (result) {
                    try {
                        JSONObject jsonObject = new JSONObject(data);
                        String ret = jsonObject.getString("result");
                        if (ret.equals("0")) {
                            JSONArray jsonArray = jsonObject.getJSONArray("items");
                            ArrayList<ItemData> dataList = new ArrayList<ItemData>();
                            for (int i = 0; i < jsonArray.length(); i++) {
                                ItemData itemData = ItemData.create(jsonArray.getJSONObject(i));
                                if (itemData != null) {
                                    dataList.add(itemData);
                                }
                            }
                            mDataList = dataList;
                            callback.didReceiveData(true);
                            return;
                        }
                    } catch(Exception e) {}
                }
                callback.didReceiveData(false);
            }
        });
        StringBuffer param = new StringBuffer();
        param.append("command=getItem");
        httpManager.execute(Constants.ServerRootUrl, "POST", param.toString());
    }

    public interface ItemRequesterCallback {
        void didReceiveData(boolean result);
    }

    public ArrayList<ItemData> getDataList() {
        return mDataList;
    }

    public ItemData query(String itemId) {

        for (int i = 0; i < mDataList.size(); i++) {
            if (mDataList.get(i).itemId.equals(itemId)) {
                return mDataList.get(i);
            }
        }
        return null;
    }

    public String queryId(String itemName) {

        for (int i = 0; i < mDataList.size(); i++) {
            if (mDataList.get(i).name.equals(itemName)) {
                return mDataList.get(i).itemId;
            }
        }
        return null;
    }
}

File: iOS/Icchi/Icchi/Http/Requester/ItemRequester.swift
//
//  ItemRequester.swift
//  Icchi
//
//  Created by oonaka on 2018/02/19.
//  Copyright © 2018年 Leapfrog-Inc. All rights reserved.
//

import Foundation

class ItemRequester {
    
    // レスポンスデータ型
    private struct fetchResult : Codable {
        var result:Bool
        var items:[ItemData]?
        private enum CodingKeys: String, CodingKey {
            case result
            case items
        }
        init(from decoder: Decoder) throws {
            let values = try decoder.container(keyedBy: CodingKeys.self)
            result = NSString(string:try values.decode(String.self, forKey: .result)).isEqual(to: "0")
            items = try values.decode([ItemData].self, forKey: .items)
        }
    }
    // アイテムデータ
    struct ItemData : Codable {
        var itemId:String? = nil
        var name:String? = nil
        var kana:String? = nil
    }
    
    // レスポンスデータ型
    private struct createResult : Codable {
        var result:Bool
        var itemId:String?
        private enum CodingKeys: String, CodingKey {
            case result
            case itemId
        }
        init(from decoder: Decoder) throws {
            let values = try decoder.container(keyedBy: CodingKeys.self)
            result = NSString(string:try values.decode(String.self, forKey: .result)).isEqual(to: "0")
            itemId = try values.decode(String.self, forKey: .itemId)
        }
    }
    
    
    // シングルトン
    static let sharedManager = ItemRequester()
    private init() {
    }
    
    // アイテムデータ配列
    fileprivate(set) var mDataList:[ItemData] = []
    
    
    /** データ取得 */
    func fetch(completion: @escaping ((Bool) -> ())) {
        
        // リクエストデータ作成
        let params = ["command": "getItem"]
        // リクエスト実施
        HttpManager.post(url: Constants.ServerApiUrl, params: params) { (result:Bool, data:Data?) in
            
            // レスポンスチェック
            guard result, let data = data else {
                // 失敗
                completion(false)
                return
            }
            
            do {
                // レスポンス取得
                let result = try JSONDecoder().decode(fetchResult.self, from: data)
                
                // アイテムデータ配列を登録
                if let users = result.items {
                    self.mDataList = users
                }
                
                // 成功
                completion(result.result)
            }
            catch {
                // 失敗
                completion(false)
            }
        }
    }
    
    /** データ取得 */
    func query(_ itemId:String) -> ItemData? {
        return mDataList.first(where: { return $0.itemId == itemId})
    }
    
    /** データID取得 */
    func queryId(_ itemName:String) -> String? {
        let data = mDataList.first(where: { return $0.name == itemName})
        return data?.itemId
    }
    
    /** データ作成 */
    func creteItem(_ itemName:String, completion: @escaping ((Bool, String?) -> ())) {
        
        // リクエストデータ作成
        let params = [
            "command": "createItem",
            "itemName": "itemName"
        ]
        // リクエスト実施
        HttpManager.post(url: Constants.ServerApiUrl, params: params) { (result:Bool, data:Data?) in
            
            // レスポンスチェック
            guard result, let data = data else {
                // 失敗
                completion(false, nil)
                return
            }
            
            do {
                // レスポンス取得
                let result = try JSONDecoder().decode(createResult.self, from: data)
                
                // 成功
                completion(result.result, result.itemId)
            }
            catch {
                // 失敗
                completion(false, nil)
            }
        }
    }
    
}
